src/eval.c

```diff
@@ -385,10 +385,15 @@ uint64_t evalGetCommandFlags(client *c, uint64_t cmd_flags) {
     int evalsha = c->cmd->proc == evalShaCommand || c->cmd->proc == evalShaRoCommand;
     if (evalsha && sdslen(c->argv[1]->ptr) != 40)
         return cmd_flags;
-    evalCalcFunctionName(evalsha, c->argv[1]->ptr, funcname);
-    char *lua_cur_script = funcname + 2;
-    dictEntry *de = dictFind(lctx.lua_scripts, lua_cur_script);
+    dictEntry *de = NULL;
     uint64_t script_flags;
+    // We only need to call the expensive function evalCalcFunctionName
+    // if we're handling EVALSHA related commands
+    if (evalsha){
+        evalCalcFunctionName(evalsha, c->argv[1]->ptr, funcname);
+        char *lua_cur_script = funcname + 2;
+        de = dictFind(lctx.lua_scripts, lua_cur_script);
+    }
```

Comment:
the idea behind calling `evalCalcFunctionName` in the original code was to be able to take the flags from the dict rather than re-parse them (i.e. save the call to `evalExtractShebangFlags`).

also, the current fix, doesn't avoid calling evalCalcFunctionName twice in case of EVALSHA.
maybe we better cache the output of both evalCalcFunctionName and evalExtractShebangFlags somewhere, so we avoid repeating them while processing the same command?

i'm not certain where's the right place to do that, maybe inside the client struct and reset it in `resetClient`?
@MeirShpilraien do you have any other suggestion?

Reply:
> the idea behind calling evalCalcFunctionName in the original code was to be able to take the flags from the dict rather than re-parse them (i.e. save the call to evalExtractShebangFlags).

Makes sense but EVAL will never make use of the cached flags so Indeed we don't need this for EVAL correct?

WRT:
> maybe we better cache the output of both evalCalcFunctionName and evalExtractShebangFlags somewhere, so we avoid repeating them while processing the same command?

If we can do that we would remove this wasteful computation! So if possible let's do that.

Reply:
@oranagra I remember we discussed the caching option and decided not to do it because it was problematic. I guess now that we see it can improve performance then we should. I will take another look at the code to refresh my memory and I might be able to suggest better.

Reply:
i've added a commit that replaces the previous approach with one that caches the funcname.
looks like it's pointless to cache the script_flags, since after the first EVAL, they'll already be in the dictionary, so it'll only help reducing repeated calls to evalExtractShebangFlags if this is the first time we see that script.

but maybe we can cache the dictEntry in server.lua_scripts instead? this can save us the repeated dictFind effort (and we can extract the function name from the key)

Reply:
made that change.
@filipecosta90 care to benchmark that for me?

Reply:
@oranagra I've updated the PR description with the new numbers. Notice that we're dropping performance after the last changes.

```
this PR iteration 1 (https://github.com/redis/redis/commit/37c20bb129c4d95ee73db0c562099b030bdbe643)

ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Evals      187842.57        10.64582        10.55900        20.86300        24.70300     23847.20 
Totals     187842.57        10.64582        10.55900        20.86300        24.70300     23847.20 

this PR (https://github.com/redis/redis/commit/27665e814be7969309de028a44ef3d6965f4d01b)

ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Evals      181413.30        11.02263        10.87900        21.50300        25.47100     23030.98 
Totals     181413.30        11.02263        10.87900        21.50300        25.47100     23030.98

```

Reply:
@filipecosta90 it makes no sense to me, maybe there's a glitch in the benchmark? something environmental changed?
the case is that we're executing the same script again and again using EVAL, so the differences are:

Your version:
* evalGetCommandFlags: calls evalExtractShebangFlags (no call to evalCalcFunctionName)
* evalGenericCommand: calls evalCalcFunctionName and dictFind
* sum: evalExtractShebangFlags + evalCalcFunctionName + dictFind

My version:
* evalGetCommandFlags: calls evalCalcFunctionName and dictFind (will not normally call and evalExtractShebangFlags)
* evalGenericCommand: calls neither
* sum: evalCalcFunctionName + dictFind

maybe it's because for this specific script, evalExtractShebangFlags is super quick, but still why would it be slower?

note that in other cases, my version has more savings, like not calling either dictFind or evalCalcFunctionName twice in case of EVALSHA.
However it doesn't reduce the need to call evalCalcFunctionName and evalExtractShebangFlags twice on the first time we see a new unique script.

Reply:
@oranagra you were right. Some glitch on the env -- all ok. 
I took the chance to also measure the improvement over EVALSHA. Notice we're still around 16% slower than 6.2.6. 

COMMAND | 6.2.6 | unstable | this PR | % Improvement vs unstable | % Drop vs 6.2.6
-- | -- | -- | -- | -- | --
EVAL | 223744 | 169767 | 187298 | 10.33% | -16.29%
EVALSHA | 264411 | 204867 | 220708 | 7.73% | -16.53%